snapshot/pkg/controller/snapshotter/snapshotter.go

```diff
@@ -371,74 +371,52 @@ func (vs *volumeSnapshotter) getPlugin(snapshotName string, snapshot *crdv1.Volu
-		}
-	} else if status == statusError {
-		return status, nil, fmt.Errorf("Failed to find snapshot %s", snapshotName)
-	}
-	bCreateSnapData = true
 	// Find snapshot by existing tags, and create VolumeSnapshotData
-
-	if bCreateSnapData {
-		snapshotDataSource, conditions, err = vs.findSnapshot(snapshotName, snapshot)
-		if err != nil {
-			return statusNew, nil, nil
-		}
-		glog.Infof("getSnapshotStatus: create VolumeSnapshotData object for VolumeSnapshot %s.", snapshotName)
-		snapshotDataObj, err := vs.createVolumeSnapshotData(snapshotName, snapshot, snapshotDataSource)
-		if err != nil {
-			return statusError, nil, err
-		}
-		if status != statusReady {
-			_, err = vs.UpdateVolumeSnapshot(snapshotName, conditions)
-			if err != nil {
-				glog.Errorf("getSnapshotStatus: Error updating volume snapshot %s: %v", snapshotName, err)
-				return statusError, nil, err
-			}
-		}
-		glog.Infof("getSnapshotStatus: bind VolumeSnapshotData to VolumeSnapshot %s.", snapshotName)
-		err = vs.bindVolumeSnapshotDataToVolumeSnapshot(snapshotName, snapshotDataObj.Metadata.Name)
-		if err != nil {
-			glog.Errorf("getSnapshotStatus: Error binding VolumeSnapshotData to VolumeSnapshot %s: %v", snapshotName, err)
-			return statusError, nil, err
-		}
-		return statusPending, snapshotDataObj, nil
+	snapshotDataSource, conditions, err = vs.findSnapshot(snapshotName, snapshot)
+	if err != nil {
+		return statusNew, nil, nil
+	}
+	// Snapshot is found. Since the status is New, it suggests that
+	// VolumeSnapshotData has not been created yet. So create VolumeSnapshotData,
+	// bind VolumeSnapshotData to VolumeSnapshot, and update VolumeSnapshot status
+	glog.Infof("updateSnapshotIfExists: create VolumeSnapshotData object for VolumeSnapshot %s.", snapshotName)
+	snapshotDataObj, err = vs.createVolumeSnapshotData(snapshotName, snapshot, snapshotDataSource, conditions)
+	if err != nil {
+		return statusError, nil, err
+	}
+	glog.Infof("updateSnapshotIfExists: update VolumeSnapshot status and bind VolumeSnapshotData to VolumeSnapshot %s.", snapshotName)
+	_, err = vs.UpdateVolumeSnapshot(snapshotName, conditions)
+	if err != nil {
+		glog.Errorf("updateSnapshotIfExists: Error updating volume snapshot %s: %v", snapshotName, err)
+		return statusError, nil, err
 	}
-	return statusError, nil, nil
+
+	return statusPending, snapshotDataObj, nil
 }
 
 // Below are the closures meant to build the functions for the GoRoutineMap operations.
 // syncSnapshot is the main controller method to decide what to do to create a snapshot.
 func (vs *volumeSnapshotter) syncSnapshot(snapshotName string, snapshot *crdv1.VolumeSnapshot) func() error {
 	return func() error {
-		status, snapshotDataObj, err := vs.getSnapshotStatus(snapshot)
+		var err error
+		var snapshotDataObj *crdv1.VolumeSnapshotData
+		status := vs.getSimplifiedSnapshotStatus(&snapshot.Status.Conditions)
+		if status == statusNew {
+			status, snapshotDataObj, err = vs.updateSnapshotIfExists(snapshot)
```

Comment:
Because the current workflow is
1. Create metadata and update in API object
2. Take a snapshot
3. Create VolumeSnapshotData and update object
4. Bind object by updating VolumeSnapshot (the status is also changed to pending at this time)

If failure happens between 3 and 4, it means that snapshot is already taken (don't even need to findSnapshot through cloud provider) since we already have a VolumeSnapshotData object created which refers to the volume snapshot. so   I think in updateSnapshotIfExists and findSnapshot function we can do the following
1. Check metadata exist or not, if no return statusNew
2. search volumeSnapshotData object which refers the the volume, if yes, update volumeSnapshot object with the snapshotData object name, returns statusPending. 
3 findSnapshot through cloud provider. If found, create snapshotData and bind it, and return statusPending. If not, return statusNew.

Reply:
but we can merge this change now, and do another round of refactoring